File: C++/VendingMachine/VendingMachine/Rails.cpp
#include "pch.h"
#include "Rails.h"
#include <map>
#include "Rail.h"

Rails::Rails(const RailsConfiguration& railsConfiguration)
{
	for (const auto& settings : railsConfiguration.GetSettings())
	{
		auto configuredRail = new Rail(settings.GetInitialInventory(), settings.GetPrice(), settings.GetLabel());

		RailMap.insert(std::make_pair(settings.GetRailCode(), configuredRail));
		RailsSummaryMap.insert(std::make_pair(settings.GetRailCode(), settings.GetLabel()));
	}
}

bool Rails::CanSelectRail(const std::string& railCode) const
{
	return RailMap.find(railCode) != RailMap.end();
}

void Rails::SelectRail(const std::string& railCode)
{
	if (CanSelectRail(railCode))
		SelectedRail = railCode;
}

Rail* Rails::GetSelectedRail()
{
	return RailMap.find(SelectedRail)->second;
}

std::map<std::string, std::string> Rails::GetRailsSummary() const
{
	return RailsSummaryMap;
}

bool Rails::HasSelectedRail() const
{
	return !SelectedRail.empty();
}

void Rails::ClearRailSelection()
{
	SelectedRail.clear();
}

File: java/VendingMachine/src/com/ravendarque/vendingMachine/rails/Rails.java
package com.ravendarque.vendingMachine.rails;

import java.util.Collections;
import java.util.HashMap;
import java.util.Map;

public class Rails {

    private static final String NO_SELECTION = "";

    private final Map<String, Rail> rails;
    private final Map<String, String> railsSummary;
    private String selectedRail = NO_SELECTION;

    public Rails(RailsConfiguration railConfiguration) {

        Map<String, Rail> configuredRails = new HashMap<>();
        Map<String, String> configuredRailsSummary = new HashMap<>();
        for (RailConfigurationSettings settings : railConfiguration.getConfigurations()) {

            final Rail rail = new Rail(
                    settings.getInitialInventory(),
                    settings.getPrice(),
                    settings.getLabel());

            configuredRails.put(settings.getRailCode(), rail);
            configuredRailsSummary.put(settings.getRailCode(), settings.getLabel());
        }

        this.rails = Collections.unmodifiableMap(configuredRails);
        this.railsSummary = Collections.unmodifiableMap(configuredRailsSummary);

    }

    public boolean canSelectRail(String railCode) {

        return rails.containsKey(railCode);
    }

    public void clearSelection() {

        selectedRail = NO_SELECTION;
    }

    public Map<String, String> getRailsSummary() {

        return railsSummary;
    }

    public Rail getSelectedRail() {

        return rails.get(selectedRail);
    }

    public boolean hasSelectedRail() {

        return !selectedRail.isEmpty();
    }

    public void selectRail(String railCode) {

        if (canSelectRail(railCode))
            selectedRail = railCode;
    }
}
